src/record_layout.zig

```diff
@@ -0,0 +1,383 @@
+const std = @import("std");
+
+const Type = @import("Type.zig");
+const Attribute = @import("Attribute.zig");
+const Compilation = @import("Compilation.zig");
+const Parser = @import("Parser.zig");
+const Record = Type.Record;
+const Field = Record.Field;
+const TypeLayout = Type.TypeLayout;
+
+const assert = std.debug.assert;
+
+// almost all the code for record layout
+// was liberally copied from this repro
+// https://github.com/mahkoh/repr-c
+// specifically the code in "repc/impl/src/" directory.
+// At the time of this writing (Q2 2022)
+// that code marked as "SPDX-License-Identifier: MIT OR Apache-2.0"
+// so is compatable with arocc's MIT licence
+
+const RecordContext = struct {
+    // Whether the record has an __attribute__((packed)) annotation.
+    attr_packed: bool,
+    // The value of #pragma pack(N) at the type level if any.
+    max_field_align_bits: ?u29,
+    // The alignment of this record.
+    aligned_bits: u29,
+    is_union: bool,
+    // The size of the record. This might not be a multiple of 8 if the record contains bit-fields.
+    // For structs, this is also the offset of the first bit after the last field.
+    size_bits: u29,
+};
+
+const int_type = Type{ .specifier = .int };
+
+pub const BITS_PER_BYTE: u29 = 8;
+
+pub fn computeLayout(ty: *Type, p: *const Parser, type_layout: *TypeLayout) void {
```

Comment:
I don't think `ty` needs to be a pointer here (or in `computeLayout`). In general passing `Type` by value is preferred. In `rawAlignOf` you can recursively call it on `ty.data.sub_type.*` for the `typeof` cases. I think that function also belongs in `Type.zig`

Reply:
fixed. Yeah, that's better.
I think there is a bunch of `fn`s in `Type.zig` that take a `*const Compilation` that could just take a `Compilation`

Reply:
Those generally all need to be pointers because `Compilation` itself is pretty heavy, and it contains allocators which usually aren't safe to copy around.